Add generic mechanism to specify constant memory locations

undingen opened the pull request:
This interface allows a program to tell the JIT about constant memory location.
I use this in the cython module to tell the JIT about constant longs.
But this should be extended to handle much more objects and types.

This does not yet remove all unnecessary code for the following example but it get's us further:
```
def fib_jit(n):
    a = 42
    b = 1000
    return a + b

```
More optimizations require patches to cpython to let it mark the `small_ints` as const and mark the return value of `MyObject_Malloc` as `noalias`.
So that things like `PyObject_Malloc(28) == Py_NotImplemented` can get optimized away (may require additional code which I already have implemented ;).

```
prime_summing:
NUM INST: 538 --> 369
NUM BB  : 107 -->  73

fib2:
NUM INST: 2136 --> 2110
NUM BB  :  428 -->  426
```


__What do you think about this approach?__
It's not very clean but I did not want to introduce knowledge about CPython into dpro.
Some of this cases can probably be handled by processing the generated bitcode and just marking constants directly in the IR. But I'm quite sure this will handle much fewer cases than this approach.

kmod commented:
Looks awesome!

kmod commented on src/jit2.cpp:
```diff
@@ -190,6 +198,17 @@ void* LLVMCompiler::compile(unique_ptr<Module> module, string funcname) {
+}
+
+bool JitConsts::isPointedToLocationConst(char* ptr, int load_size) {
+    return consts.overlaps(ptr, ptr + load_size);
```
I tried this pr out on "100 + 101" and got something funky:

```
define dso_local %struct._object.2497* @traced_1(%struct._object.2497* nocapture readnone, %struct._object.2497* nocapture readnone) #0 {
  store i64 4, i64* inttoptr (i64 9171136 to i64*), align 64, !tbaa !0
  ret %struct._object.2497* inttoptr (i64 9171136 to %struct._object.2497*)
}
```

ie it inadvertently treated the refcount as constant as well.  I dug into it and the [llvm docs](http://llvm.org/doxygen/classllvm_1_1IntervalMap.html) for IntervalMap say the ranges are inclusive, so I think we want to change these overlaps/insert calls to `(ptr, ptr + load_size - 1)'?

undingen replied:
Oh how embarrassing :blush:
thanks for catching this!

undingen commented:
So I rebased this PR (and made sure I merged in your `RuntimeValue` changes)
afterwards I added a new commit which replaces all references to globals with the address of the symbol.

With this change we compile:
```python
    a = 10
    b = 32
    return a + b
```
to:

```llvm
define dso_local %struct._object.5386* @traced_1(%struct._object.5386* nocapture readnone, %struct._object.5386* nocapture readnone) #0 {
  %3 = load i64, i64* inttoptr (i64 9156832 to i64*), align 32, !tbaa !0
  %4 = add nsw i64 %3, 1
  store i64 %4, i64* inttoptr (i64 9156832 to i64*), align 32, !tbaa !0
  %5 = load i64, i64* inttoptr (i64 9157536 to i64*), align 32, !tbaa !0
  %6 = add nsw i64 %5, 1
  store i64 %6, i64* inttoptr (i64 9157536 to i64*), align 32, !tbaa !0
  %7 = load i64, i64* inttoptr (i64 9157856 to i64*), align 32, !tbaa !0
  %8 = add i64 %7, 1
  store i64 %8, i64* inttoptr (i64 9157856 to i64*), align 32, !tbaa !0
  %9 = load i64, i64* inttoptr (i64 9156832 to i64*), align 32, !tbaa !0
  %10 = add nsw i64 %9, -1
  store i64 %10, i64* inttoptr (i64 9156832 to i64*), align 32, !tbaa !0
  %11 = icmp eq i64 %10, 0
  br i1 %11, label %12, label %13, !prof !6

12:                                               ; preds = %2
  tail call void inttoptr (i64 4484816 to void (%struct._object.5386*)*)(%struct._object.5386* nonnull inttoptr (i64 9156832 to %struct._object.5386*)) #3
  br label %13

13:                                               ; preds = %2, %12
  %14 = load i64, i64* inttoptr (i64 9157536 to i64*), align 32, !tbaa !0
  %15 = add nsw i64 %14, -1
  store i64 %15, i64* inttoptr (i64 9157536 to i64*), align 32, !tbaa !0
  %16 = icmp eq i64 %15, 0
  br i1 %16, label %17, label %18, !prof !6

17:                                               ; preds = %13
  tail call void inttoptr (i64 4484816 to void (%struct._object.5386*)*)(%struct._object.5386* nonnull inttoptr (i64 9157536 to %struct._object.5386*)) #3
  br label %18

18:                                               ; preds = %17, %13
  ret %struct._object.5386* inttoptr (i64 9157856 to %struct._object.5386*)
}
```

so only the refcounting operations are left (4484816 is long_dealloc).

__Whats your opinion on last commit?__
It definitely makes it a bit harder to read the LLVM IR...

undingen commented:
mmh so I guess GVN? or what pass should eliminate the dead stores is not smart enough to eliminate them if it sees inttoptr instructions :-(
It's reproducible with clang: https://godbolt.org/z/olkpyg gcc is smart enough to optimize it away...

kmod commented:
I guess we'll just see how bad the readability is :)  Also I think we could add an AliasAnalysis pass that adds this sort of different-constant-addresses-don't-alias support if it ends up being helpful.

One thing in the back of my mind is trace caching; I don't know if it should be a goal but I wonder if we would have to switch away from replacing globals for that to work.

kmod commented:
I added a simple Alias Analysis pass for a different reason, feel free to add this to it!  Check out DproAAResult::alias()
